rust_src/src/crypto/mod.rs

```diff
@@ -414,20 +415,30 @@ fn _secure_hash(algorithm: HashAlg, input: &[u8], hex: bool) -> LispObject {
         hash_func(input, digest_str.as_mut_slice());
     }
     if hex {
-        make_digest_string(digest_str.as_mut_slice(), digest_size);
+        hexify_digest_string(digest_str.as_mut_slice(), digest_size);
     }
     digest
 }
 
-fn make_digest_string(buffer: &mut [u8], len: usize) {
+/// To avoid a copy, buffer is both the source and the destination of
+/// this transformation. Buffer must contain len bytes of data and
+/// 2*len bytes of space for the final hex string.
+fn hexify_digest_string(buffer: &mut [u8], len: usize) {
     static hexdigit: [u8; 16] = *b"0123456789abcdef";
+    debug_assert!(buffer.len() == 2 * len, "buffer must be long enough to hold 2*len hex digits");
     for i in (0..len).rev() {
         let v = buffer[i];
         buffer[2 * i] = hexdigit[(v >> 4) as usize];
         buffer[2 * i + 1] = hexdigit[(v & 0xf) as usize];
     }
 }
 
+// For the following hash functions, the caller must ensure that the
+// destination buffer is at least long enough to hold the
+// digest. Additionall, the caller may have been asked to return a hex
+// string, in which case dest_buf will be twice as long as the digest.
+// Thus, these functions are unsafe.
```

Comment:
Ah, sorry, I misremembered that `copy_from_slice` requires same length. But that's easy: make it `dest_buf[..output.len()].copy_from_slice(output)`.